### PR TITLE
Improve update release channels

### PR DESCRIPTION
**Pull request by zhuangbiaowei**
## Summary
- add stable and beta release channels for codewhale update
- skip checksum and binary downloads when the local version is already current or newer
- add semver-based version comparison and focused tests

## Tests
- cargo fmt
- cargo test -p codewhale-cli

<details open><summary><h3>Greptile Summary</h3></summary>

This PR introduces stable and beta release channels for `codewhale update`, adds semver-based version comparison to skip redundant downloads, and includes focused unit tests for the new logic.

- `ReleaseChannel`, `FetchedRelease`, and `ReleaseSource` types are added to `update.rs`; `run_update` now accepts a `beta: bool` flag wired to a new `--beta` CLI argument in `lib.rs`.
- `fetch_latest_beta_release_from_url` scans the `/releases?per_page=100` list for the first tag containing \"beta\"; `update_is_needed` uses `semver` to guard against redundant downloads and cross-channel downgrades.
- The mirror-source path correctly emits a warning when `--beta` is used, addressing a concern from the previous review thread.
</details>

<details open><summary><h3>Confidence Score: 5/5</h3></summary>

Safe to merge; the channel-switching and downgrade-prevention logic is well-tested and the previous review concerns have been addressed.

The core version-comparison logic in update_is_needed handles the key edge cases (stable-to-beta channel switch, cross-version downgrade prevention, RC vs beta disambiguation) and every case is backed by a unit test. The remaining observations are quality improvements rather than present defects.

crates/cli/src/update.rs — specifically the stable-channel branch of update_is_needed and the mirror-source early-exit path.
</details>

<details open><summary><h3>Important Files Changed</h3></summary>

| Filename | Overview |
|----------|----------|
| crates/cli/src/update.rs | Core update logic: adds ReleaseChannel, FetchedRelease, update_is_needed, fetch_latest_beta_release_from_url; previous-thread issues are addressed; stable channel has no prerelease guard; mirror path skips the version check |
| crates/cli/src/lib.rs | Adds UpdateArgs struct with --beta flag and plumbs it into run_update; CLI parsing test added; straightforward change |
| Cargo.toml | Adds semver 1.0.28 as workspace dependency |
| crates/cli/Cargo.toml | Picks up semver from workspace |
| Cargo.lock | Lock file updated for semver dependency |

</details>

</details>

<details open><summary><h3>Flowchart</h3></summary>

```mermaid
%%{init: {'theme': 'neutral'}}%%
flowchart TD
    A[codewhale update] -->|--beta flag| B{ReleaseChannel}
    B -->|Stable| C[fetch /releases/latest]
    B -->|Beta| D[fetch /releases?per_page=100]
    C --> E{Mirror env set?}
    D --> E
    E -->|Yes: Mirror| F[release_from_mirror_base_url]
    F --> G{channel == Beta?}
    G -->|Yes| H[Print mirror warning]
    G -->|No| I[Proceed to download]
    H --> I
    E -->|No: GitHub| J[update_is_needed?]
    J -->|No: already current| K[Exit: no download needed]
    J -->|Yes| I
    I --> L[Download checksum manifest]
    L --> M[Download + verify binaries]
    M --> N[Replace binaries atomically]
```
</details>

<a href="https://chatgpt.com/codex/deeplink?prompt=IMPORTANT%3A%20Work%20in%20the%20repository%20%22hmbown%2Fcodewhale%22%20on%20the%20existing%20branch%20%22improve-update-channels%22.%20Checkout%20that%20branch%20%E2%80%94%20do%20NOT%20create%20a%20new%20branch%20or%20open%20a%20new%20PR.%20Push%20your%20changes%20to%20%22improve-update-channels%22.%0A%0AFix%20the%20following%203%20code%20review%20issues.%20Work%20through%20them%20one%20at%20a%20time%2C%20proposing%20concise%20fixes.%0A%0A---%0A%0A%23%23%23%20Issue%201%20of%203%0Acrates%2Fcli%2Fsrc%2Fupdate.rs%3A498%0AThe%20stable%20channel's%20%60update_is_needed%60%20check%20has%20no%20guard%20against%20installing%20a%20prerelease.%20In%20semver%2C%20%600.8.45%20%3C%200.9.0-beta.1%60%20evaluates%20to%20%60true%60%2C%20so%20if%20a%20mirror%20URL%20or%20an%20unexpected%20GitHub%20response%20ever%20returns%20a%20prerelease%20tag%20as%20the%20%22latest%20stable%22%2C%20this%20function%20would%20approve%20installing%20it.%20The%20test%20on%20line%20982%20explicitly%20documents%20this%20behaviour.%20A%20simple%20%60!latest.pre.is_empty%28%29%60%20check%20can%20prevent%20it.%0A%0A%60%60%60suggestion%0A%20%20%20%20%20%20%20%20ReleaseChannel%3A%3AStable%20%3D%3E%20Ok%28current%20%3C%20latest%20%26%26%20latest.pre.is_empty%28%29%29%2C%0A%60%60%60%0A%0A%23%23%23%20Issue%202%20of%203%0Acrates%2Fcli%2Fsrc%2Fupdate.rs%3A42-52%0A**Mirror%20source%20skips%20version%20check%20for%20stable%20too**%0A%0AThe%20%60if%2Felse%20if%60%20structure%20means%20%60update_is_needed%60%20is%20never%20called%20for%20mirror%20sources%20on%20any%20channel%20%E2%80%94%20not%20just%20%60--beta%60.%20A%20stable%20user%20whose%20mirror%20happens%20to%20serve%20the%20same%20version%20they%20already%20have%20will%20re-download%20and%20reinstall%20every%20time%20they%20run%20%60codewhale%20update%60.%20The%20PR%20description%20says%20%22skip%20checksum%20and%20binary%20downloads%20when%20the%20local%20version%20is%20already%20current%20or%20newer%22%2C%20but%20that%20improvement%20only%20applies%20to%20GitHub%20sources.%0A%0A%23%23%23%20Issue%203%20of%203%0Acrates%2Fcli%2Fsrc%2Fupdate.rs%3A1137-1145%0AThe%20new%20%60fetch_latest_beta_release_from_url%60%20test%20checks%20the%20%60Accept%60%20header%20but%20not%20the%20%60User-Agent%60%2C%20unlike%20the%20existing%20%60fetch_latest_release_from_url%60%20test.%20Adding%20the%20assertion%20keeps%20the%20two%20tests%20consistent%20and%20would%20catch%20a%20regression%20if%20%60update_http_client%60%20stopped%20setting%20the%20user-agent.%0A%0A%60%60%60suggestion%0A%20%20%20%20%20%20%20%20assert!%28%0A%20%20%20%20%20%20%20%20%20%20%20%20request_lower.contains%28%22accept%3A%20application%2Fvnd.github%2Bjson%22%29%2C%0A%20%20%20%20%20%20%20%20%20%20%20%20%22got%20%7Brequest%3A%3F%7D%22%0A%20%20%20%20%20%20%20%20%29%3B%0A%20%20%20%20%20%20%20%20assert!%28%0A%20%20%20%20%20%20%20%20%20%20%20%20request_lower.contains%28%22user-agent%3A%20codewhale-updater%22%29%2C%0A%20%20%20%20%20%20%20%20%20%20%20%20%22got%20%7Brequest%3A%3F%7D%22%0A%20%20%20%20%20%20%20%20%29%3B%0A%20%20%20%20%20%20%20%20handle.join%28%29.expect%28%22test%20server%20thread%22%29%3B%0A%20%20%20%20%7D%0A%0A%20%20%20%20%23%5Btest%5D%0A%20%20%20%20fn%20fetch_latest_beta_release_from_url_reports_missing_beta%28%29%20%7B%0A%60%60%60%0A%0A"><picture><source media="(prefers-color-scheme: dark)" srcset="https://greptile-static-assets.s3.amazonaws.com/badges/FixAllInCodexDark.svg?v=3"><source media="(prefers-color-scheme: light)" srcset="https://greptile-static-assets.s3.amazonaws.com/badges/FixAllInCodex.svg?v=3">Fix All in Codex</picture></a> <a href="https://app.greptile.com/ide/claude-code?prompt=Fix%20the%20following%203%20code%20review%20issues.%20Work%20through%20them%20one%20at%20a%20time%2C%20proposing%20concise%20fixes.%0A%0A---%0A%0A%23%23%23%20Issue%201%20of%203%0Acrates%2Fcli%2Fsrc%2Fupdate.rs%3A498%0AThe%20stable%20channel's%20%60update_is_needed%60%20check%20has%20no%20guard%20against%20installing%20a%20prerelease.%20In%20semver%2C%20%600.8.45%20%3C%200.9.0-beta.1%60%20evaluates%20to%20%60true%60%2C%20so%20if%20a%20mirror%20URL%20or%20an%20unexpected%20GitHub%20response%20ever%20returns%20a%20prerelease%20tag%20as%20the%20%22latest%20stable%22%2C%20this%20function%20would%20approve%20installing%20it.%20The%20test%20on%20line%20982%20explicitly%20documents%20this%20behaviour.%20A%20simple%20%60!latest.pre.is_empty%28%29%60%20check%20can%20prevent%20it.%0A%0A%60%60%60suggestion%0A%20%20%20%20%20%20%20%20ReleaseChannel%3A%3AStable%20%3D%3E%20Ok%28current%20%3C%20latest%20%26%26%20latest.pre.is_empty%28%29%29%2C%0A%60%60%60%0A%0A%23%23%23%20Issue%202%20of%203%0Acrates%2Fcli%2Fsrc%2Fupdate.rs%3A42-52%0A**Mirror%20source%20skips%20version%20check%20for%20stable%20too**%0A%0AThe%20%60if%2Felse%20if%60%20structure%20means%20%60update_is_needed%60%20is%20never%20called%20for%20mirror%20sources%20on%20any%20channel%20%E2%80%94%20not%20just%20%60--beta%60.%20A%20stable%20user%20whose%20mirror%20happens%20to%20serve%20the%20same%20version%20they%20already%20have%20will%20re-download%20and%20reinstall%20every%20time%20they%20run%20%60codewhale%20update%60.%20The%20PR%20description%20says%20%22skip%20checksum%20and%20binary%20downloads%20when%20the%20local%20version%20is%20already%20current%20or%20newer%22%2C%20but%20that%20improvement%20only%20applies%20to%20GitHub%20sources.%0A%0A%23%23%23%20Issue%203%20of%203%0Acrates%2Fcli%2Fsrc%2Fupdate.rs%3A1137-1145%0AThe%20new%20%60fetch_latest_beta_release_from_url%60%20test%20checks%20the%20%60Accept%60%20header%20but%20not%20the%20%60User-Agent%60%2C%20unlike%20the%20existing%20%60fetch_latest_release_from_url%60%20test.%20Adding%20the%20assertion%20keeps%20the%20two%20tests%20consistent%20and%20would%20catch%20a%20regression%20if%20%60update_http_client%60%20stopped%20setting%20the%20user-agent.%0A%0A%60%60%60suggestion%0A%20%20%20%20%20%20%20%20assert!%28%0A%20%20%20%20%20%20%20%20%20%20%20%20request_lower.contains%28%22accept%3A%20application%2Fvnd.github%2Bjson%22%29%2C%0A%20%20%20%20%20%20%20%20%20%20%20%20%22got%20%7Brequest%3A%3F%7D%22%0A%20%20%20%20%20%20%20%20%29%3B%0A%20%20%20%20%20%20%20%20assert!%28%0A%20%20%20%20%20%20%20%20%20%20%20%20request_lower.contains%28%22user-agent%3A%20codewhale-updater%22%29%2C%0A%20%20%20%20%20%20%20%20%20%20%20%20%22got%20%7Brequest%3A%3F%7D%22%0A%20%20%20%20%20%20%20%20%29%3B%0A%20%20%20%20%20%20%20%20handle.join%28%29.expect%28%22test%20server%20thread%22%29%3B%0A%20%20%20%20%7D%0A%0A%20%20%20%20%23%5Btest%5D%0A%20%20%20%20fn%20fetch_latest_beta_release_from_url_reports_missing_beta%28%29%20%7B%0A%60%60%60%0A%0A&repo=hmbown%2Fcodewhale&pr=2145&platform=github"><picture><source media="(prefers-color-scheme: dark)" srcset="https://greptile-static-assets.s3.amazonaws.com/badges/FixAllInClaudeDark.svg?v=3"><source media="(prefers-color-scheme: light)" srcset="https://greptile-static-assets.s3.amazonaws.com/badges/FixAllInClaude.svg?v=3">Fix All in Claude Code</picture></a> <a href="https://app.greptile.com/api/ide/cursor?prompt=Fix%20the%20following%203%20code%20review%20issues.%20Work%20through%20them%20one%20at%20a%20time%2C%20proposing%20concise%20fixes.%0A%0A---%0A%0A%23%23%23%20Issue%201%20of%203%0Acrates%2Fcli%2Fsrc%2Fupdate.rs%3A498%0AThe%20stable%20channel's%20%60update_is_needed%60%20check%20has%20no%20guard%20against%20installing%20a%20prerelease.%20In%20semver%2C%20%600.8.45%20%3C%200.9.0-beta.1%60%20evaluates%20to%20%60true%60%2C%20so%20if%20a%20mirror%20URL%20or%20an%20unexpected%20GitHub%20response%20ever%20returns%20a%20prerelease%20tag%20as%20the%20%22latest%20stable%22%2C%20this%20function%20would%20approve%20installing%20it.%20The%20test%20on%20line%20982%20explicitly%20documents%20this%20behaviour.%20A%20simple%20%60!latest.pre.is_empty%28%29%60%20check%20can%20prevent%20it.%0A%0A%60%60%60suggestion%0A%20%20%20%20%20%20%20%20ReleaseChannel%3A%3AStable%20%3D%3E%20Ok%28current%20%3C%20latest%20%26%26%20latest.pre.is_empty%28%29%29%2C%0A%60%60%60%0A%0A%23%23%23%20Issue%202%20of%203%0Acrates%2Fcli%2Fsrc%2Fupdate.rs%3A42-52%0A**Mirror%20source%20skips%20version%20check%20for%20stable%20too**%0A%0AThe%20%60if%2Felse%20if%60%20structure%20means%20%60update_is_needed%60%20is%20never%20called%20for%20mirror%20sources%20on%20any%20channel%20%E2%80%94%20not%20just%20%60--beta%60.%20A%20stable%20user%20whose%20mirror%20happens%20to%20serve%20the%20same%20version%20they%20already%20have%20will%20re-download%20and%20reinstall%20every%20time%20they%20run%20%60codewhale%20update%60.%20The%20PR%20description%20says%20%22skip%20checksum%20and%20binary%20downloads%20when%20the%20local%20version%20is%20already%20current%20or%20newer%22%2C%20but%20that%20improvement%20only%20applies%20to%20GitHub%20sources.%0A%0A%23%23%23%20Issue%203%20of%203%0Acrates%2Fcli%2Fsrc%2Fupdate.rs%3A1137-1145%0AThe%20new%20%60fetch_latest_beta_release_from_url%60%20test%20checks%20the%20%60Accept%60%20header%20but%20not%20the%20%60User-Agent%60%2C%20unlike%20the%20existing%20%60fetch_latest_release_from_url%60%20test.%20Adding%20the%20assertion%20keeps%20the%20two%20tests%20consistent%20and%20would%20catch%20a%20regression%20if%20%60update_http_client%60%20stopped%20setting%20the%20user-agent.%0A%0A%60%60%60suggestion%0A%20%20%20%20%20%20%20%20assert!%28%0A%20%20%20%20%20%20%20%20%20%20%20%20request_lower.contains%28%22accept%3A%20application%2Fvnd.github%2Bjson%22%29%2C%0A%20%20%20%20%20%20%20%20%20%20%20%20%22got%20%7Brequest%3A%3F%7D%22%0A%20%20%20%20%20%20%20%20%29%3B%0A%20%20%20%20%20%20%20%20assert!%28%0A%20%20%20%20%20%20%20%20%20%20%20%20request_lower.contains%28%22user-agent%3A%20codewhale-updater%22%29%2C%0A%20%20%20%20%20%20%20%20%20%20%20%20%22got%20%7Brequest%3A%3F%7D%22%0A%20%20%20%20%20%20%20%20%29%3B%0A%20%20%20%20%20%20%20%20handle.join%28%29.expect%28%22test%20server%20thread%22%29%3B%0A%20%20%20%20%7D%0A%0A%20%20%20%20%23%5Btest%5D%0A%20%20%20%20fn%20fetch_latest_beta_release_from_url_reports_missing_beta%28%29%20%7B%0A%60%60%60%0A%0A&pr=2145&platform=github"><picture><source media="(prefers-color-scheme: dark)" srcset="https://greptile-static-assets.s3.amazonaws.com/badges/FixAllInCursorDark.svg?v=3"><source media="(prefers-color-scheme: light)" srcset="https://greptile-static-assets.s3.amazonaws.com/badges/FixAllInCursor.svg?v=3">Fix All in Cursor</picture></a>

<sub>Reviews (3): Last reviewed commit: ["Prevent beta update downgrades"](https://github.com/hmbown/codewhale/commit/6b6101739e8ad1d0aed7bc0128639a7a7ac4c03e) | [Re-trigger Greptile](https://app.greptile.com/api/retrigger?id=33847822)</sub>

### CODE REVIEW

File: crates/cli/src/update.rs [hunk 1]
```diff
@@ -21,18 +22,26 @@ const LEGACY_UPDATE_VERSION_ENV: &str = "DEEPSEEK_VERSION";
...
-    let release = fetch_latest_release().with_context(update_network_fallback_hint)?;
+    let release = fetch_latest_release(channel).with_context(update_network_fallback_hint)?;
     let latest_tag = &release.tag_name;
-    println!("Latest release: {latest_tag}");
+    println!("Latest {} release: {latest_tag}", channel.label());
+
+    if !update_is_needed(current_version, latest_tag)? {
+        println!("Already up to date; no download needed.");
+        return Ok(());
+    }
```
[the remark this answers is not in the document]
Comment (zhuangbiaowei) [reply crates/cli/src/update.rs]: resolved
